src/test/java/org/elasticsearch/transport/local/ContextLocalTransportTests.java

```diff
@@ -0,0 +1,450 @@
+            return actionGet.getSuggest();
+        } else {
+            SuggestRequestBuilder builder = client().prepareSuggest().setIndices(queryIndex).putHeader(randomHeaderKey, randomHeaderValue);
+            if (suggestText != null) {
+                builder.setSuggestText(suggestText);
+            }
+            for (SuggestBuilder.SuggestionBuilder<?> suggestion : suggestions) {
+                builder.addSuggestion(suggestion);
+            }
+
+            SuggestResponse actionGet = builder.execute().actionGet();
+            assertThat(Arrays.toString(actionGet.getShardFailures()), actionGet.getFailedShards(), equalTo(expectShardsFailed));
+            if (expectShardsFailed > 0) {
+                throw new SearchPhaseExecutionException("suggest", "Suggest execution failed", new ShardSearchFailure[0]);
+            }
+            return actionGet.getSuggest();
+        }
+    }
+
+    private void assertGetRequestsContainHeaders() {
+        assertGetRequestsContainHeaders(this.lookupIndex);
+    }
+
+    private void assertGetRequestsContainHeaders(String index) {
+        for (GetRequest request : requests) {
+            if (!request.index().equals(index)) {
+                continue;
+            }
+
+            assertThat(String.format(Locale.ROOT, "Expected header %s to be in get requests for index %s", randomHeaderKey, index), request.hasHeader(randomHeaderKey), is(true));
+            assertThat(request.getHeader(randomHeaderKey).toString(), is(randomHeaderValue));
+        }
+    }
+
+    public static class ActionLoggingPlugin extends AbstractPlugin {
+
+        @Override
+        public String name() {
+            return "test-action-logging";
+        }
+
+        @Override
+        public String description() {
+            return "Test action logging";
+        }
+
+        @Override
+        public Collection<Class<? extends Module>> modules() {
+            return ImmutableList.of(ActionLoggingModule.class);
+        }
+    }
+
+    public static class ActionLoggingModule extends AbstractModule implements PreProcessModule {
+
+
+        @Override
+        protected void configure() {
+            bind(LoggingFilter.class).asEagerSingleton();
+        }
+
+        @Override
+        public void processModule(Module module) {
+            if (module instanceof ActionModule) {
+                ((ActionModule)module).registerFilter(LoggingFilter.class);
+            }
+        }
+    }
+
+    public static class LoggingFilter extends ActionFilter.Simple {
+
+        @Inject
+        public LoggingFilter(Settings settings) {
+            super(settings);
+        }
+
+        @Override
+        public int order() {
+            return 999;
+        }
+
+        @Override
+        protected boolean apply(String action, ActionRequest request, ActionListener listener) {
+//            logger.info("### action [{}], request [{}]", action, request.getClass().getName());
+            if (request instanceof GetRequest) {
```

Comment:
need to try this out for headers, by setting some header in the client() and transportClient() settings (not sure if we propagate those). This will not work at all with the context, because that only lives on the local node and is not copied.. unless we fill the context for every request manually, but then it doesnt make a lot of sense to check if it is configured.